INTER-233-CardComponent-공용 컴포넌트 Card 구현

mulgko opened the pull request:
## PR 제목
Inter 233 card component

### 반영 브랜치
Inter 233 card component -> dev

### PR 타입(하나 이상의 PR 타입을 선택해주세요)
- [x] 기능 추가
- [ ] 기능 삭제
- [ ] 버그 수정
- [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트
- [ ] 코드 리펙토링

### 작업 내용
- main, favorite, edit, detail, check 카드 생성

### 테스트 결과 (선택)
<img width="628" alt="image" src="https://github.com/user-attachments/assets/49cbf183-67a1-49c8-8d95-9c81fba1836b" />

https://github.com/user-attachments/assets/6880b377-e2b6-4c35-8ed0-b82a94596a62

### 리뷰 요구사항
- 컴포넌트들 이름 추천 좀 해주세용

mulgko committed:
feat: card component 추가
- check
- detail
- main
- edit

mulgko committed:
refactor: 코드 리팩토링

mulgko committed:
feat: card bg category에 따라서 변경

mulgko committed:
feat: 기능 추가
- 체크박스 추가
-  아이콘 추가
- image 추가
- 동적 props 추가

mulgko committed:
refactor: 코드 리팩토링

GiToon10100011 commented on src/app/components/card/constants.ts:
정말 죄송하지만.. 이렇게 사용하는건 너무 사용이 불편하고, 재사용성을 위한 컴포넌트 제작에 전혀 알맞지 않은것 같습니다. 이렇게 하면 고정적인 데이터로만 카드를 사용할 수 밖에 없게 되며, 유동적으로 컴포넌트 사용이 불가하여 사용성이 저하됩니다. 모든 데이터는 애초에 db로 부터 받아오기 때문에, 해당 자료는 추후에 전혀 필요가 없어질 예정입니다.. 애초에 카드는 List 컴포넌트와 함께 사용되어 map으로 돌려서 만들어질텐데.. 그럼 차라리 ERD에 맞게 타입 인터페이스를 정의해서 props를 통해 해당 데이터들을 들어갈 창구를 카드 컴포넌트에 만들어줬어야합니다. 카드 컴포넌트는 어디까지나 틀이며, 안에 들어갈 요소를 어느정도의 패딩으로 가져갈거고, 어느 사이즈로 나올건지 지정해주는건데, 데이터까지 고정시켜놓으면 그냥 페이지에서 작업하는 것과 동일한 결과가 발생합니다.

GiToon10100011 replied:
제가 그래서 제안하는 수정방안은 먼저 type을 props로 받고, 해당 type에 맞는 스타일을 적용시킨 후, 데이터는 db에서 받아온 질문 배열에 들어가있는 속성들에 맞게 인터페이스를 정의해서 각 속성을 사용할 창구를 JSX로 만들어놓는 것입니다.

GiToon10100011 replied:
그럼 일단 constants.ts는 아예 지우시고, type.ts도 그냥 ICardProps하나만 만들어서 type이랑 children정도만 받으면 될 것 같습니다.. 사실 제가 의도했던건 각 카드는 어차피 각 페이지에 사용용도가 국한되므로, 전역적으로 사용되는 컴포넌트가 아니라, 해당 페이지의 로컬 컴포넌트로 제작을 하는 것이었는데, 이미 다 작업을 한걸 또 전부 삭제하긴 너무 아까우므로 일단은 전부 지우고, 각 컴포넌트마다 어떤식으로 데이터를 넣을것인지만 정해놓고, 로컬 컴포넌트로 해당 컴포넌트들을 나중에 페이지 작업시 옮기는것을 제안해봅니다.

mulgko replied:
넵! 해당 내용으로 수정하겠습니다 감사합니다 :)

GiToon10100011 commented on src/app/components/card/constants.ts:
```diff
@@ -0,0 +1,212 @@
+import { ICardDataProps } from "./card.type";
+
+export const mainCardData: ICardDataProps[] = [
```
예를 들어보자면, IMainCardProps라는 인터페이스로, 필요한것만 사용한다고 가졍한다면,

```ts
title: string;
description: string;
image: string;
author: string;
category: string;
level: "easy" | "medium" | "hard"
tags: ITag[] (Tag라는 별도의 인터페이스를 선언)
hearts: number;
isLiked: boolean;
```

일단 대충 요런식으로? 정의하면 되지않을까 싶습니다. 이후, 예를들어 tags같은 경우는 복합컴포넌트로 해당 컴포넌트 안에 List를 다시 써서 type을 tags로 하고, Tag 컴포넌트를 다시 맵함수로 돌려서 전달하여 렌더링하면 됩니다.

GiToon10100011 commented on src/hooks/card/useResponsiveSize.ts:
사실 해당 훅함수 또한 필요가 없어지는게, 각 페이지별 로컬 컴포넌트로 작업을하면 각 컴포넌트에 맞게 테일윈드로 스타일만 만들어주면 되기 때문에 mobile여부는 그냥 단순히 현재 만들어둔 useViewport만 사용하면 될 것 같습니다..

GiToon10100011 commented on src/app/components/card/card.variants.ts:
만약 수정을 하면 해당 파일에서는 각 타입에 맞는 패딩이나 너비 높이 와 같은 일종의 틀 역할을 해줄 스타일링만 하면 됩니다.

GiToon10100011 replied:
또한 FormData제외, any타입을 사용하는것은 금지입니다.

GiToon10100011 reviewed (changes requested):
총 요약을 하자면, 결국 만들어야할 파일은 Card.tsx, card.type.ts, card.variants.ts, index.ts 외에는 없으며, 각 역할군에 따른 카드 컴포넌트는 각 페이지의 로컬 컴포넌트로 마이그레이션을 해야합니다. (우선 만들어둔 각 카드별 컴포넌트들은 보존, 추후 페이지 작업 시 마이그레이션 진행)또한 Card.tsx는 일종의 틀과같은 역할만하는거로, children을 사용하여 어떤 요소든 받게 하고, type을 통해 variants에 맞는 스타일을 적용시키기만 하면 됩니다. 그렇게 하면 로컬 컴포넌트는 다음과 같은 방법으로 만들어질 겁니다. 

```ts

// app/main/components/mainCard/MainCard.tsx
const MainCard = () => {
  <Card type="mainCard">
    //...들어갈 내용
  </MainCard>
}

//사용 시: 
// app/main/page.tsx
  <List type="main">
    {data.map((el) => <MainCard key={key} {...el} />)}
  </List>

```

애써 열심히 만든걸 전부 수정하게 만들어 죄송합니다. 
이해가 안되시면 얼마든지 질문 주시고 작업하실때 현재 작업하시는거에 의문이나 제대로 하는게 맞는건지 잘 모르시겠다면 얼마든지 상담해주세요

GiToon10100011 commented on src/app/components/card/card.variants.ts:
```diff
@@ -0,0 +1,78 @@
+  variants: {
+    size: {
+      "410": {
+        container: "w-[414px] h-[414px] py-4 px-4",
```
추후 수정하실때 arbitrary 클래스는 사용하지 마시고 w-103 h-103과 같은 방식으로 수정 부탁드립니다.

mulgko replied:
로컬에서 수정완료하였습니다

GiToon10100011 commented on src/app/components/card/cards/DetailCard.tsx:
```diff
@@ -0,0 +1,68 @@
+                    {problem.question}
+                  </p>
+                ))}
+                <p className="font-medium-12 md:font-mediut-12 text-sub-text">
```
mediut 오타 났습니다

GiToon10100011 commented on src/app/components/card/cards/EditCard.tsx:
```diff
@@ -0,0 +1,98 @@
+                  {problem.question}
+                </p>
+              ))}
+              <p className="font-medium-12 md:font-mediut-12 text-sub-text">
```
여기도 medium으로 수정부탁드립니다

mulgko committed:
refactor: 컴포넌트 코드 변경
- caards 삭제 (main, favorite, edit, detail, check)
- hooks(cards_Responsive) 삭제
- varaints 타입 정의
- type.ts 파일 재정의

mulgko committed:
refactor: 리팩토링

GiToon10100011 commented:
LGTM!
